Fixes #6007 - Allowing to hide smart class parameters

orrabin opened the pull request.

orrabin committed:
Fixes #6007 - Allowing to hide smart class parameters

tbrisker commented on app/views/api/v2/smart_class_parameters/main.json.rabl:
```diff
@@ -2,5 +2,5 @@ object @smart_class_parameter
 extends "api/v2/smart_class_parameters/base"
 
-attributes :description, :override, :parameter_type, :default_value, :use_puppet_default, :required, :validator_type, :validator_rule,
+attributes :description, :override, :parameter_type, :default_value, :hidden_value, :use_puppet_default, :required, :validator_type, :validator_rule,
```
why not remove `:default_value` and `:hidden_value` and only show `:safe_value`?

orrabin replied:
I wanted to keep the same behavior as globals, maybe both should be changed?

domcleal replied:
I think :hidden_value is definitely needed, when you consider a tool like Hammer is going to want to know and change a parameter's hidden_value attribute.

Whether the :default_value should be hidden or not in the API response, I'm unsure.  I think putting the onus on the API consumer to hide it is OK, since there might be cases (like in our UI) where having the non-hidden value available is still desired.

domcleal replied:
Is this hidden_value field actually going to contain `*****`?  I had assumed (and haven't checked the global parameter field) that it would contain true/false, i.e. the value of the `hidden_value` column.

I can't see a lot of use of sending asterisks to an API user, but we really ought to be providing them with the value of the boolean so they can toggle it on and off.

orrabin replied:
I agree but I didn't want to remove is from global parameters, I'll replace with `hidden_value?` here and add it to global parameters

tbrisker commented on app/helpers/lookup_keys_helper.rb:
```diff
@@ -78,7 +78,8 @@ def hostgroup_key_with_diagnostic(hostgroup, key)
     original_value = key.value_before_type_cast value
     diagnostic_helper = popover('', _("<b>Description:</b> %{desc}<br><b>Type:</b> %{type}<br> <b>Matcher:</b> %{matcher}") % { :desc => key.description, :type => key.key_type, :matcher => origin}, :data => { :placement => 'top' })
-    parameter_value_content("value_#{key.key}", original_value, :popover => diagnostic_helper)
+    value_to_display = key.hidden_value? ? key.hidden_value : original_value
```
Why not use `:safe_value` here?

orrabin replied:
`safe_value` will return key.value if it's not hidden and `original_value` returns `key.value_before_type_cast value`

domcleal commented on app/assets/javascripts/lookup_keys.js:
```diff
@@ -282,3 +282,24 @@ function build_match() {
+    target = $('<textarea/>').attr({class: 'form-control', id: source.id, name: source.name, placeholder: 'Value', rows: 1, value: $(source).val()});
+  }
+  $(source).replaceWith(target);
+}
```
Can you check your editor config please?  It should keep an EOL at the end of the file.

domcleal commented on db/migrate/20151025120534_add_hidden_value_to_lookup_key.rb:
```diff
@@ -0,0 +1,5 @@
+  def change
+    add_column :lookup_keys, :hidden_value, :boolean, :default => false
+  end
+end
```
[fixed, EOL] ditto

domcleal commented on app/views/lookup_keys/_fields.html.erb:
```diff
@@ -17,16 +17,26 @@
+                   :size => "col-md-1", :label_size => "col-md-3", :table_field => true,
                    :onchange => 'toggleUsePuppetDefaultValue(this, "default_value")',
                    :disabled => (is_param && !f.object.override)) if is_param %>
+    <%= checkbox_f(f, :hidden_value, :label => _('Hidden value'),
```
[fixed] Can you add a help_inline to this checkbox, like the others?

domcleal commented on app/views/common_parameters/_puppetclass_parameter.html.erb:
```diff
@@ -20,11 +20,15 @@ end
-            <%= fullscreen_input %>
+          <div class="input-group col-md-12">
+            <% if f.object.hidden_value? %>
+              <%= f.password_field(:value, :class => 'form-control', :value => f.object.value) %>
```
I think this needs a :disabled property in case use_puppet_default is true.

Edit: or indeed, authorisation?

domcleal commented:
1) hidden_value needs adding to the API controller parameters docs.

2) Nitpicky, but the strikethrough of a wildcard value inherited on the host edit form looks more like it's underlined.  Is there any way to redesign that so it's clearer?

![screenshot from 2015-11-02 14-35-07](https://cloud.githubusercontent.com/assets/482089/10884277/09a7e6e0-816f-11e5-9594-3f156b52250b.png)

tbrisker commented:
@domcleal:  2. ~~*****~~  (strike-through doesn't play nice with asterisks, I suggest removing the `text-decoration: line-through;` CSS from overridden hidden params)

domcleal commented on app/controllers/api/v2/smart_class_parameters_controller.rb:
```diff
@@ -36,6 +36,7 @@ def show
         param :override, :bool, :desc => N_("Whether the smart class parameter value is managed by Foreman")
         param :description, String, :desc => N_("Description of smart class")
         param :default_value, String, :desc => N_("Value to use when there is no match")
+        param :hidden_value, :bool, :desc => N_("Value to display when the parameter should be hidden")
```
Not really a "value to display", because it's a boolean that controls _if_ the parameter should be hidden.  Perhaps: "Hide the parameter value [in the UI?] when enabled"

orrabin replied:
`hidden_value` is not the boolean it's "*****" this is how it's displayed for global parameters, I can add the boolean for both global parameters and here but I was trying to keep the same behavior.

domcleal replied:
Oh heck, I think I see what you mean.  `#hidden_value` is `*****` but `#hidden_value?` is the boolean attribute, right?

Remember that this here is the documentation for incoming parameters on create/update, _not_ response parameters, so if a user specifies `hidden_value` then it would be there to update the boolean attribute.  I'll put another comment inline for the response/rabl part.

domcleal commented on app/controllers/api/v2/smart_variables_controller.rb:
```diff
@@ -28,6 +28,7 @@ def show
           param :variable, String, :required => true, :desc => N_("Name of variable")
           param :puppetclass_id, :number, :desc => N_("Puppet class ID")
           param :default_value, String, :desc => N_("Default value of variable")
+          param :hidden_value, :bool, :desc => N_("Value to display when the parameter should be hidden")
```
ditto

domcleal commented on app/views/lookup_keys/_fields.html.erb:
```diff
@@ -17,16 +17,28 @@
+    <%= checkbox_f(f, :hidden_value, :label => _('Hidden value'),
+                   :class => 'hidden_value_textarea_switch', :onchange => 'turn_default_value_and_lookup_values_switch(this)',
+                   :checked => f.object.hidden_value?,
+                   :help_inline => popover("", _("Hide all values for this parameter")),
```
Nitpick, please end this string in a full stop (`.`) to match the others.

domcleal commented on app/views/lookup_keys/_fields.html.erb:
```diff
@@ -17,16 +17,28 @@
-    <%= checkbox_f(f, :use_puppet_default, :label => _('Use Puppet default'), :size => "col-md-8",
+    <% if f.object.hidden_value? %>
+        <%= password_f f, :default_value, :size => "col-md-8", :no_label => false, :value => f.object.default_value_before_type_cast,
+                       :disabled => (is_param && (!f.object.override || f.object.use_puppet_default)), :fullscreen => true %>
```
I don't think this should have :fullscreen, since the host edit partial doesn't.

tbrisker replied:
This is for consistency of the layout when flipping between hidden and not-hidden, and also to save the need to add and remove the fullscreen button when doing so.

domcleal commented:
> @domcleal: 2. ****\* (strike-through doesn't play nice with asterisks, I suggest removing the text-decoration: line-through; CSS from overridden hidden params)

Yeah, that's probably the only sensible thing to do.

orrabin commented:
@domcleal I fixed your comments, some of the asterisks are still stricken through because there isn't a single selector to get them all but they will be fixed when: https://github.com/theforeman/foreman/pull/2858 is merged.
Should I fix all the cases or just leave the strike through for all of them for now? I don't want part of them to be with the line and others without.

domcleal commented on app/views/api/v2/override_values/base.json.rabl:
```diff
@@ -1,3 +1,3 @@
 object @override_value
 
-attributes :id, :match, :value, :use_puppet_default
+attributes :id, :match, :value, :hidden_value, :use_puppet_default
```
Not sure this is any help, apart from a little bit of consistency with common parameters.  I'd suggest leaving it out so we can delete it in future?

orrabin replied:
I'm removing it from here and `main.json.rabl`

domcleal commented on app/views/lookup_keys/_value.html.erb:
```diff
@@ -10,9 +10,13 @@
-                      :disabled => f.object.use_puppet_default,
-                      :placeholder => _("Value") %>
+      <% if f.object.hidden_value? %>
+          <%= password_field_tag 'value', f.object.value, :class => 'form-control', :disabled => f.object.use_puppet_default %>
```
Nitpick, this has four-space indentation between the if statement and password_field_tag, but should be two.

domcleal commented:
> some of the asterisks are still stricken through because there isn't a single selector to get them all but they will be fixed when: #2858 is merged.
> Should I fix all the cases or just leave the strike through for all of them for now? I don't want part of them to be with the line and others without.

I don't mind, entirely up to you.  I'd suggest doing whichever will make conflict resolution, or tidying up after 2858 is merged (if it's after - shall we wait in this PR?) easier.  e.g. if it's easier to leave the strikethrough in place now and then fix it in a couple of lines later, do that.

orrabin commented:
@domcleal easier to keep everything with strikethrough since #2858 removed the strikethrough from everything since the override is done inline.
There is no need to tidy up after, it will be resolved on it's own so they can be merged in any order.

domcleal commented:
Sure, okay.  Could you look at my other inline comments please?

orrabin commented:
@domcleal thanks, I fixed the comments and rebased.

domcleal commented:
1) There's a bug in the smart class edit form with the full screen button:
1. enable hidden value on a parameter and save it
2. open the form again, untick the hidden value field
3. press full screen button, the editor is blank

It looks to me like the JS on the button is accessing a span for the caps lock warning message (for the textbox) instead of the input field.

2) I'm a little confused about the behaviour of the host edit form.  With a global parameter we always had the hide checkbox on the form (because it was an attr on the parameter itself) so you could toggle it on/off and view the unhidden value.

Because the hide attr is on the key here, it's not possible to override a smart class parameter but also view the value temporarily.  That might be a problem for any class parameter containing complex YAML etc as you won't know the formatting to be able to edit it.  It'll work OK for a plain text parameter (e.g. a password) though.  Do you think this is a problem?  Should we have a way to show the unhidden parameter in an editor temporarily?

orrabin commented:
@domcleal comment 1 is fixed, also rebased and added a helper `hidden_value_field` for all of the `value` fields that needed hiding (`default_value` was a little different).

I agree with the problem in comment 2 but adding another checkbox to the table in hosts edit page might be too crowded. Maybe this should go in and we should wait for feedback? I can add it and send both screenshots to the list but I'm not sure that would get as much notice.

domcleal commented:
Would it make sense to show the unhidden value in the full screen view on the host edit form?  So you're unable to see the hidden value in the single-line text box, but can by pressing that button.

domcleal commented:
Could you also now rebase this please?

domcleal commented:
> I agree with the problem in comment 2 but adding another checkbox to the table in hosts edit page might be too crowded

Looks like you've added another button throughout, right?

domcleal commented on app/assets/javascripts/lookup_keys.js:
```diff
@@ -278,3 +278,34 @@ function build_match() {
   });
 }
 
+function  toggle_lookupkey_hidden(checkbox) {
```
Nitpick, extra space after `function`

domcleal commented on app/assets/javascripts/lookup_keys.js:
```diff
@@ -278,3 +278,34 @@ function build_match() {
+  toggle_value_hidden(default_value);
+  lookup_values.each(function () {
+    toggle_value_hidden($(this));
+  })
```
missing semicolon

domcleal commented on app/assets/javascripts/lookup_keys.js:
```diff
@@ -278,3 +278,34 @@ function build_match() {
+    $(target).replaceWith($('<input/>').attr(_.extend({ type: 'password' }, attrs)));
+  }
+  else {
+    $(target).replaceWith($('<textarea/>').attr(_.extend({ placeholder: 'Value', rows: 1 }, attrs)));
```
`Value` looks untranslated

domcleal commented on app/assets/javascripts/lookup_keys.js:
```diff
@@ -278,3 +278,34 @@ function build_match() {
+  target.closest('tr').find('.set_hidden_value').prop('checked', shown);
+  if (shown) {
+    $(target).replaceWith($('<input/>').attr(_.extend({ type: 'password' }, attrs)));
+  }
```
Nitpick, keep `} else {` on the same line.

orrabin commented:
@domcleal @tbrisker added a button to this and it looks better then the checkbox, it also replaces the checkbox for global parameters.
For smart class parameters it still only changes the UI and doesn't change if the key is actually hidden or not, we have been debating if it should change the key or not.

domcleal commented:
> For smart class parameters it still only changes the UI and doesn't change if the key is actually hidden or not, we have been debating if it should change the key or not.

If not, then it may be worth disabling the button when the key isn't a hidden value to avoid people misunderstanding where it's set (i.e. if they were to hide the value in the host#edit UI, but it not to actually be saved that way).

domcleal commented:
There appears to be a bug on the host#edit page when you toggle between hidden value and override buttons on a smart class param.

a. toggle hidden button on, then off on a smart class parameter (without host override)
b. click Override button next to parameter
c. only the "Value" placeholder is shown in the text box

orrabin commented:
@domcleal the hide button now controls the lookup_key and I fixed the bug

domcleal commented:
Thanks, though an integration test is failing, please have a look.

orrabin commented:
@domcleal I fixed the failing test and rebased

domcleal commented on app/models/lookup_value.rb:
```diff
@@ -37,6 +38,11 @@ def value=(val)
 
+  def hidden_value=(val)
+    lookup_key.hidden_value = val
+    lookup_key.save unless lookup_key.new_record?
```
This is causing updates for every (overridden?) lookup key it seems on a host#edit save.  I'm seeing lots of UPDATE SQL queries in the log when saving which are amounting to no-op changes:

<pre>
2015-12-01T11:34:21 [sql] [D]    (0.2ms)  UPDATE "lookup_keys" SET "updated_at" = '2015-12-01 11:34:21.532348', "default_value" = '--- my default!
 | ...
 | ' WHERE "lookup_keys"."type" IN ('PuppetclassLookupKey') AND "lookup_keys"."id" = 110
</pre>


Where it is actually changing the default value, it does seem to work but rewrites the default value each time.

A couple extra points about this:

a) Should this use an autosave association, perhaps from lookup value to lookup key?  I'm not sure if that might impact other parts of the app though, e.g. the Puppet class or lookup key edit pages.
b) This shouldn't save until the LookupValue itself is saved.
c) Will failures of lookup_key.save be detected?

orrabin replied:
autosave caused problems in other places which is why this was added, I'll change it save the lookup key only before the lookup value is saved and only if `hidden_value` was updated.

unorthodoxgeek replied:
c. lookup_key is saved in the DB, so it _should_ already be valid. there is no validation on the hidden_value field on it, so this shouldn't be a problem...

unorthodoxgeek commented on app/helpers/common_parameters_helper.rb:
```diff
@@ -16,17 +16,36 @@ def parameter_value_field(value)
+                    :name => options[:name].to_s,
+                    :disabled => options[:disabled] }
+    if lookup_key.present? && lookup_key.hidden_value?
+      text_area = password_field_tag(id, value, option_hash)
```
I'd name it field or something, as it can be a password field, in which case, the name "text_area" is, well, odd.

unorthodoxgeek commented on app/helpers/common_parameters_helper.rb:
```diff
@@ -16,17 +16,36 @@ def parameter_value_field(value)
+
+  def hidden_value_field(f, field, value, disabled, options = {})
+    if f.object.hidden_value?
+      text_field = f.password_field(field, :disabled => disabled, :value => value, :class => 'form-control')
```
ditto

domcleal replied:
This comment's unresolved.

domcleal replied:
(It was about the naming of the variable, "text_field".)

unorthodoxgeek commented on app/models/lookup_keys/lookup_key.rb:
```diff
@@ -69,7 +75,7 @@ def self.inherited(child)
     :use_puppet_default,
     :validator_type, :validator_rule,
-    :variable, :variable_type
+    :variable, :variable_type,
```
what is this comma?

unorthodoxgeek commented on app/views/api/v2/common_parameters/main.json.rabl:
```diff
@@ -2,4 +2,4 @@ object @common_parameter
 extends "api/v2/common_parameters/base"
 
-attributes :created_at, :updated_at, :hidden_value
+attributes :created_at, :updated_at, :hidden_value?
```
why the question mark? does it produce a hidden_value? key? if so, this isn't what should be here. I'd do

``` ruby
node(:hidden_value){|i| i.hidden_value?}
```

not sure if that's the correct rabl syntax. point is, we need this to work like it did before.

same for both files below this one.

orrabin replied:
This is not for lookup_keys, the (global) parameters showed "*****" instead of actually showing if they are hidden or not, this fixes it.

domcleal replied:
I'd suggest we keep both (I think this was the case at some point), so we have the new field (the boolean ending in `?`) and the old, useless field.

domcleal commented:
There's a test failure, could you check it please?

orrabin commented:
@domcleal it's a result of changing the `hidden_value=` into a `before_save` https://github.com/theforeman/foreman/pull/2866/files#diff-eed2703a4841c9402fcb9258af61f058R16
I've been working on it and having trouble because `lookup_value_attributes=` in host_common doesn't call any of the callbacks for lookup_value.
I can call `save` explicitly in `lookup_value_attributes=` but it might be something we are missing that does exist in the original method created by `accepts_nested_attributes_for`.
I'm still checking it but right now I have 2 easy options:
1) add `lookup_value.save` in `lookup_value_attributes=` 
2) decide to not save the value of `hidden_value` from host/hostgroup edit form (so I'd also disable the button on all lookup_keys that are not hidden)
I'd appreciate your opinion or any idea what else can be done to call the callbacks.

domcleal commented:
> I've been working on it and having trouble because lookup_value_attributes= in host_common doesn't call any of the callbacks for lookup_value.

I think what should happen is that defining accepts_nested.. for lookup values will enable autosave on the Host->LookupValue association.  That ought to add a before_save callback or similar to save the LVs when the Host is saved, so I'm not sure why you'd not see a callback on LV.  You shouldn't ever see a save directly from `lookup_value_attributes=` (it should only call #attributes=).

Regarding ways to save the lookup key from the host, rather than adding the .save inside the setter (which still concerns me, as the save hasn't been requested), could you add a before_save in host_common rather than lookup_value which calls save?

Or define an association on host_common to LookupKey through LookupValues, then enable autosave on that?

> 2) decide to not save the value of hidden_value from host/hostgroup edit form (so I'd also disable the button on all lookup_keys that are not hidden)

I also don't mind this, it's perhaps what I would have done.  I expect we can get this working though.

orrabin commented:
@domcleal none of the options I tried worked for all cases and it's probably best to edit the hidden option from the key anyway as everything else that belongs to the key is edited in one place.
Using the same icon on both when one saves the hide/unhide choice and the other just shows the value temporarily is confusing so there are two different icons to help distinguish between the behaviors.
Also, the additional info no longer shows the inherited value if the key is hidden (it shows *****).

domcleal commented:
The icon's not as good, but I think I agree with the reasoning, thanks.

orrabin commented:
@domcleal I fixed the comment I missed about the field name and added both `hidden_value` and `hidden_value?` to all of the rabl files I'm changing here.

domcleal commented on app/helpers/common_parameters_helper.rb:
```diff
@@ -16,17 +16,38 @@ def parameter_value_field(value)
+    if f.object.hidden_value?
+      field = f.password_field(field, :disabled => disabled, :value => value, :class => 'form-control')
+    else
+      field = f.text_area(field, options.merge(:disabled => disabled,
```
In changing the name here it's reassigning/overlapping with the existing `field` variable, I'd avoid doing that to keep the code clearer.

domcleal commented on app/helpers/common_parameters_helper.rb:
```diff
@@ -16,17 +16,38 @@ def parameter_value_field(value)
+      field = f.password_field(field, :disabled => disabled, :value => value, :class => 'form-control')
+    else
+      field = f.text_area(field, options.merge(:disabled => disabled,
+                                                    :class => "form-control no-stretch",
```
indentation's now off here

orrabin commented:
@domcleal fixed the indentation and name

domcleal commented:
Merged as e4ae4a6, thanks @orrabin.
